### PR TITLE
Use fallback approach for nested softmax (+backward) and bmm

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #85441

Follow up would be to use masked_softmax

### CODE REVIEW

**Review by drisspg — COMMENTED** [review]
Looks good, and thanks for updating the softmax backward formula. Since we are doing these changes for performance reasons could you add some profiles of differences before and after? Even just timeit would be interesting for reasonable sized nested tensors. I think running the tutorial on master and then running the tutorial on this branch before and after and profiling the times would be great to include in this PR.

**Comment by mikaylagawarecki** [comment]
```
from unittest import load_tests
import torch
import numpy as np
import time
from itertools import product

torch.manual_seed(5)
print("Using cuda: ", torch.cuda.is_available())

def benchmark_torch_function(iters, f, *args, **kwargs):
    f(*args, **kwargs)
    if torch.cuda.is_available():
        torch.cuda.synchronize()
        start_event = torch.cuda.Event(enable_timing=True)
        end_event = torch.cuda.Event(enable_timing=True)
        start_event.record()
    else:
        t0 = time.time()
    for _ in range(iters):
        f(*args, **kwargs)
    if torch.cuda.is_available():
        end_event.record()
        torch.cuda.synchronize()
        return start_event.elapsed_time(end_event)
    else:
        return (time.time() - t0)

def fn1(nt1, nt2):
    y = nt1.bmm(nt2)

def zipf_sentence_lengths(alpha: float, batch_size: int, *, np_seed=0) -> np.ndarray:
    # sync generated lengths across runs
    np.random.seed(np_seed)
    sentence_lengths = np.empty(batch_size, dtype=int)
    for ibatch in range(batch_size):
        sentence_lengths[ibatch] = 1
        word = np.random.zipf(alpha)
        while word != 3 and word != 386 and word != 858:
            sentence_lengths[ibatch] += 1
            word = np.random.zipf(alpha)
    print(sentence_lengths)
    return sentence_lengths

alpha = 1.2
N = 512
E_q, E_k, E_v, E_total, E_out = 512, 512, 512, 512, 512
nheads = 8
E_head = E_total // nheads
iters = 50
dtype = torch.float64
# sentence_lengths = zipf_sentence_lengths(alpha, N)

# In tutorial we use matmul here attn_output = attn_weights.bmm(value)
# (N, nheads, L_t, L_s) x (N, nheads, L_s, E_head) -> (N, nheads, L_t, E_head)

# for padded version bmm is used attn_output = attn_weights.bmm(value)
# (N * nheads, L_t, L_s) x (N * nheads, L_s, E_head) -> (N * nheads, L_t, E_head)
# attn_output = attn_weights.bmm(value)

# L_t and L_s are jagged and are the sentence_lengths above

    # generate 10 random zipf law data and run each for 50 iters
    # return avg(avg_100_iters_i)
for i in range(10):
    print("i: ", i)
    sentence_lengths = zipf_sentence_lengths(alpha, N, np_seed=i)
    L_s = L_t = sentence_lengths
    max_len = sentence_lengths.max()
    fake_attn_weights = torch.randn(N, max_len, max_len, dtype=dtype)
    fake_value = torch.randn(N, max_len, E_v, dtype=dtype)

    nested_fake_attn_weights_size = torch.empty(N, 2, dtype=torch.long)
    nested_fake_attn_weights_size[:, 0] = torch.from_numpy(L_s)
    nested_fake_attn_weights_size[:, 1] = torch.from_numpy(L_t)
    nested_fake_attn_values_size = torch.empty(N, 2, dtype=torch.long)
    nested_fake_attn_values_size[:, 0] = torch.from_numpy(L_t)
    nested_fake_attn_values_size[:, 1] = E_v

    nested_fake_attn_weights = torch._nested_from_padded(fake_attn_weights, nested_fake_attn_weights_size)
    nested_fake_value = torch._nested_from_padded(fake_value, nested_fake_attn_values_size)

    bmm_times = []
    bmm_times.append(benchmark_torch_function(iters, fn1, nested_fake_attn_weights, nested_fake_value) / iters)
bmm_times = np.array(bmm_times)
avg_time = np.mean(bmm_times)
sd = np.std(bmm_times)
print(f'{dtype} bmm() avg_time: {avg_time:3.4f} \u00B1 {sd:3.4f}')
print('')
```
